### PR TITLE
fix: withdraw background command notices the model already read

### PR DESCRIPTION
**Pull request by snimu**
Reading a background command's result should end the story. Today it does not: a `bash()` command usually exits before the model looks at the handle, so the completion notice is queued while the model is already reading the result in the next cell. The model then gets a second, redundant message announcing work it just inspected, and that message starts an extra turn.

A result read now withdraws the notice. The kernel records the read on the handle: a read before the notice is dispatched cancels it outright, and a read after the host accepted it sends one `bash.consumed` request. The host drops the matching notice while it is still queued. Delivered notices stay: they are no longer clearable, so nothing is retracted after the model saw a notice, and an unread completion still wakes an idle session exactly as before.

One rule decides what counts as a read, for every reader (`await handle`, `poll()`, `output()`, `tail()`):

- The read must happen while a cell is executing, because the value, and anything the reader prints, ride that cell's turn. A detached watcher polling the handle between turns leaves the notice in place - that notice is the only wake-up an idle session gets.
- An awaiting reader must also be one the live cell waits for. A discarded or detached wrapper task (`create_task`, `as_completed`) awaits a value the model never sees.

The live cell comes from the kernel's active-cell state, not the cell contextvar, which detached tasks keep long after their cell ended. Reads of a still-running command's buffer never count.

Notices are matched on pid and command, and one read withdraws the earliest match only: a read belongs to one handle, so a pid reused for an identical command cannot take a second, unread notice down with the first.

The user-facing label for these messages becomes "Background command finished" instead of "Shell message received", in the queue preview and the collapsed transcript line. The model-facing first line keeps the bracket grammar `[bash-done pid:... exit:...]`.

Ordering: the kernel only arms the withdrawal after the host has accepted the notice, so a withdrawal can never arrive before the notice it withdraws. A withdrawal for a notice the host does not hold is a no-op. A host that does not know `bash.consumed` answers with an error reply, which the kernel ignores, so an older host keeps today's behavior.

Stacked on #2115 (`snimu/synthetic-message-grammar`), which owns the same `messages.ts` seam. Review the top four commits only; rebase onto main if #2115 merges first.

## LOC

- `prime-agent-runtime/src/rlm/bash.py`: +85 / -2
- `prime-agent-runtime/src/rlm/repl.py`: +12 / -0
- `packages/coding-agent/src/core/agent-session.ts`: +35 / -0
- `packages/coding-agent/src/core/rlm-runtime.ts`: +22 / -0
- `packages/coding-agent/src/core/messages.ts`, `prompts/rlm.ts`, `docs/rlm.md`: 4 lines
- tests (`test_repl.py`, `async-bash-completion`, `2053`, `2068`): +178 / -11
- changelog fragment: +1

Total: 337 insertions, 15 deletions across 12 files.

## Validation

- `npm run check` clean before every commit.
- Full `packages/coding-agent` vitest suite in a Prime sandbox: 4994 passed, 59 failed in 7 files (`extensions-discovery`, `extensions-runner`, `extensions-input-event`, `config`, `resource-loader`, `tools`, `4603`), all in the known environmental set and identical to the pre-change list.
- Full `prime-agent-runtime` python suite: OK on CPython 3.11 (sandbox and local) and on 3.13, so the consumption rule does not depend on asyncio wrapper internals.
- Real-kernel end-to-end pin in `2053-background-bash-passivation.test.ts`: the notice queues while the model is busy, the next cell reads the result, and no message is delivered.
- Pins for the negative cases too: a detached read between turns keeps the notice, a discarded or detached awaiter keeps it, a read after delivery retracts nothing, and one read withdraws one notice.

Fixes RES-1341

> [!NOTE]
> ### Withdraw unread background bash completion notices when the model reads the result
> - When a live REPL cell reads a finished background bash handle before the completion notice is delivered, the notice is suppressed entirely; if the notice was already accepted by the host, a `bash.consumed` notification retracts the queued notice
> - `BashHandle` tracks per-handle consumption state. Readers (`output`, `tail`, `poll`, `__await__`) record consumption only when the active cell transitively owns the handle; detached reads between cells do not consume
> - The TypeScript `AgentSession._withdrawAsyncBashCompletionNotice` helper finds and cancels the first matching queued notice by pid and command, then emits a queue update. Already-delivered notices are untouched
> - Renames the async bash completion preview label to `Background command finished` in [messages.ts](https://github.com/PrimeIntellect-ai/prime-agent/pull/2159/files#diff-73eea4d8b225402c51131d2c2e34f8c5c0a555902f25a75555185ccd6f090553)\n- Risk: `_schedule_background_completion_notice` in [bash.py](https://github.com/PrimeIntellect-ai/prime-agent/pull/2159/files#diff-f2d79e9ee0bfb6f41372db8ae482096e4f76ba1d03444cfe4b90cd0b0b599287) changes notice delivery timing — if a consumed callback races with host acceptance, exactly one `bash.consumed` is sent; rejected notices still use the stderr fallback. `_note_result_consumed` requires both a complete result and a live cell owner, so reads during cell transitions produce no notification
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized b68c14c.</sup>
> 
>

---

> [!NOTE]
> **Medium Risk**
> Changes steering queue behavior and host/kernel protocol (`bash.consumed`); incorrect live-cell detection could drop needed wake-ups or leave duplicate notices, but tests pin the edge cases.
> 
> **Overview**
> When a live REPL cell reads a finished `bash()` handle (`await`, `poll()`, `output()`, or `tail()`), the agent no longer gets a redundant **Background command finished** steering message or an extra turn.
> 
> The kernel tracks one-shot **result consumed** state on `BashHandle`, using `repl.active_cell_task()` so only reads during an executing cell count—detached watchers between turns still leave the notice for idle wake-ups. After the host accepts `bash.completed`, a later qualifying read fires **`bash.consumed`**; **`AgentSession._withdrawAsyncBashCompletionNotice`** cancels the earliest matching clearable queued notice (matched on **pid + command**). Notices already delivered to the model are not retracted.
> 
> User-facing queue/transcript preview text changes from **Shell message received** to **Background command finished**; docs and RLM prompts describe the new withdrawal rule.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit b68c14ce8bef283ef3ab2321e8e25f15d8867fd5. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: prime-agent-runtime/src/rlm/bash.py [hunk 1]
```diff
@@ -364,14 +376,17 @@ def running(self) -> bool:
 
     def output(self) -> str:
         self._released = True
+        self._note_result_consumed()
```
[the remark this answers is not in the document]
Comment (snimu) [reply prime-agent-runtime/src/rlm/bash.py:381]: Fixed in c2b33d675. You were right about the failure mode: a detached watcher's post-completion `poll()`/`output()` withdrew the notice, and since background output only ships with the next execution, an idle session lost its only wake-up.

The gate is not the creating cell though - a read from any *later* cell must still withdraw, which is the point of the change. The rule is now one predicate for every reader: a read counts only while a cell is executing, and an awaiting reader must also be one that live cell waits for. The live cell comes from the kernel's active-cell state, not the cell contextvar, because detached tasks keep their spawning cell's context after that cell ends. New pin: `test_detached_read_between_turns_keeps_bash_completion`.

File: packages/coding-agent/src/core/agent-session.ts [hunk 2]
```diff
@@ -4907,6 +4908,31 @@ export class AgentSession {
...
+	private _isAsyncBashCompletionActionFor(
+		action: QueuedSessionAction,
+		details: { pid: number; command: string },
+	): boolean {
+		if (action.payload.kind !== "turn") return false;
+		const message = primaryDeliveryRecord(action).message;
+		if (message.role !== "custom" || message.customType !== ASYNC_BASH_COMPLETION_CUSTOM_TYPE) return false;
+		// pids are reused across handles, so the command has to match too.
+		const completion = message.details as AsyncBashCompletionDetails | undefined;
+		return completion?.pid === details.pid && completion.command === details.command;
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/core/agent-session.ts:4939]: Narrowed in 78ff02080: the withdrawal now cancels the earliest matching notice only, instead of every action with the same key. A read belongs to one handle, and on a pid-reuse collision the earliest match is that older handle's notice, so a second unread notice survives. Pinned in the 2068 suite with two identical queued notices.

Keeping pid+command as the key: the 32-hex activity id rides the bash-activity display event, not the `bash.completed` payload or its message details, so matching on it means a new field on the kernel/host wire plus both-direction compat handling - disproportionate to a window that needs pid wraparound, an identical command string, and two notices pending at once, with one missing notice as the worst case.

File: prime-agent-runtime/src/rlm/bash.py [hunk 3]
```diff
@@ -633,6 +650,31 @@ def _add_done_callback(self, callback: Callable[[], None]) -> None:
...
+        must also be one the live cell waits for, because a discarded or detached
+        wrapper task awaits a value the model never sees.
+        """
+        if not self._done.is_set():
+            return
+        owner = _live_cell_owner()
+        if owner is None:
+            return
+        if awaiter is not None and not _creating_cell_waits_for(owner, awaiter):
+            return
```
[the remark this answers is not in the document]
Comment (snimu) [reply prime-agent-runtime/src/rlm/bash.py:669]: Declining this one. Consumption now requires a cell to be executing, and a live cell means the session is not idle: the model is inside a turn whose result it receives, and anything the reader printed arrives with that turn as background output. So the stated consequence ("an idle session is not woken") cannot happen - the idle case is the between-turns read, which is exactly what the live-cell gate protects and what `test_detached_read_between_turns_keeps_bash_completion` pins.

The missing "wait check" for `output()`/`poll()`/`tail()` is deliberate, not an oversight: a synchronous reader has no awaiter task to resolve ownership against, and contextvars are copied into detached tasks (and absent in threads), so there is nothing that distinguishes a detached synchronous reader from the model's own in-cell code. The residual case is a model-authored watcher that reads the result into a variable during someone else's turn and prints nothing; that read is the model's own program consuming the value, and treating it as a read is the rule this PR ships.

**Review by sethkarten — APPROVED** [review]
